modyn/trainer_server/internal/grpc/trainer_server_grpc_servicer.py

```diff
@@ -222,6 +246,47 @@ def get_training_status(
         cleaned_kwargs = {k: v for k, v in response_kwargs_finished.items() if v is not None}
         return TrainingStatusResponse(**cleaned_kwargs)  # type: ignore[arg-type]
 
+    def get_values_from_queues(
+        self, training_id: int
+    ) -> Tuple[Optional[bool], Optional[int], Optional[int], Optional[int], Optional[int]]:
+        was_training = self._training_process_dict[training_id].was_training
+        if was_training:
+            training_num_batches, training_num_samples, is_training = self.get_status_training(training_id, timeout=15)
+            if not is_training:
+                # clean the training queue (epoch is finished)
+                self.clean_training_queue(training_id)
+
+            # read the second queue only if the first one is empty
+            if training_num_batches is None:
+                downsampling_num_batches, downsampling_num_samples, is_training = self.get_status_downsampling(
+                    training_id, timeout=15
+                )
+            else:
+                downsampling_num_batches, downsampling_num_samples = None, None
+        else:
+            downsampling_num_batches, downsampling_num_samples, is_training = self.get_status_downsampling(
+                training_id, timeout=15
+            )
+            if is_training:
+                # clean the downsampling queue (downsampling is ended)
+                self.clean_downsampling_queue(training_id)
+
+            # read the second queue only if the first one is empty
+            if downsampling_num_batches is None:
+                training_num_batches, training_num_samples, is_training = self.get_status_training(
+                    training_id, timeout=15
+                )
+            else:
+                training_num_batches, training_num_samples = None, None
+        self._training_process_dict[training_id].was_training = is_training
```

Comment:
Exactly. The problem is managing the timeouts of the two queues. Was_training is used to maintain a state, so that we read with a higher timeout from the queue in which we expect to find something useful (the same one we read from in the previous iteration)